applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_nodal_displacement_response_function.cpp

```diff
@@ -15,40 +15,43 @@
 
 // Project includes
 #include "adjoint_nodal_displacement_response_function.h"
+#include "processes/find_elements_neighbours_process.h"
 
 namespace Kratos
 {
     AdjointNodalDisplacementResponseFunction::AdjointNodalDisplacementResponseFunction(ModelPart& rModelPart, Parameters ResponseSettings)
     : AdjointStructuralResponseFunction(rModelPart, ResponseSettings)
     {
-        // Get id of node where a displacement should be traced
-        const int id_traced_node = ResponseSettings["traced_node_id"].GetInt();
-
-        // Get the corresponding dof to the displacement which should be traced
-        // by this response function e.g. DISPLACEMENT_X, ROTATION_X,...
+        mResponsePartName = ResponseSettings["response_part_name"].GetString();
+        mResponseDirection = ResponseSettings["direction"].GetVector();
         mTracedDofLabel = ResponseSettings["traced_dof"].GetString();
 
-        // Get pointer to traced node
-        mpTracedNode = rModelPart.pGetNode(id_traced_node);
+        if ( norm_2( mResponseDirection ) > 1.0e-7 ){
+            mResponseDirection /= norm_2( mResponseDirection );
+        } else {
+            KRATOS_ERROR << "AdjointNodalDisplacementResponseFunction: 'response_direction' must not have a norm of 0.0." << std::endl;
+        }
 
         // Check if variable for traced dof is valid
-        if( !( KratosComponents<Variable<double>>::Has(mTracedDofLabel)) )
-            KRATOS_ERROR << "Specified traced DOF is not available. Specified DOF: " << mTracedDofLabel << std::endl;
-        else
-        {
-            const DoubleVariableType& r_traced_dof =
-                KratosComponents<DoubleVariableType>::Get(mTracedDofLabel);
-            KRATOS_ERROR_IF_NOT( mpTracedNode->SolutionStepsDataHas(r_traced_dof) )
-                << "Specified DOF is not available at traced node." << std::endl;
+        if( !( KratosComponents<ArrayVariableType>::Has(mTracedDofLabel)) ) {
+            KRATOS_ERROR << "AdjointNodalDisplacementResponseFunction: Specified traced DOF is not available. Specified DOF: " << mTracedDofLabel << std::endl;
+        }
+        else {
+            ModelPart& response_part = rModelPart.GetSubModelPart(mResponsePartName);
+            const ArrayVariableType& r_traced_dof =
+                KratosComponents<ArrayVariableType>::Get(mTracedDofLabel);
+            for(auto& node_i : response_part.Nodes()){
+                KRATOS_ERROR_IF_NOT( node_i.SolutionStepsDataHas(r_traced_dof) )
+                    << "AdjointNodalDisplacementResponseFunction: Specified DOF is not available at traced node." << std::endl;
+            }
```

Comment:
```suggestion
                KRATOS_ERROR_IF_NOT( node_i.HasDofFor(r_traced_dof) )
                    << "AdjointNodalDisplacementResponseFunction: Specified DOF is not available at traced node." << std::endl;
            }
```

Reply:
this will not work because the "real" dofs are `ADJOINT_DISPLACEMENT`or `ADJOINT_ROTATION` but here it shall be checked for `DISPLACEMENT`or `ROTATION`